File: SDS_알고리즘/Day03/가운데를말해요_1655.java
package SDS_알고리즘.Day03;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;

public class 가운데를말해요_1655 {
    static PriorityQueue<Integer> minHeap = new PriorityQueue<>();
    static PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
    static StringBuilder answer = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        while (N-- > 0) {
            int n = Integer.parseInt(br.readLine());
            findMid(n);
        }
        System.out.println(answer);
    }

    private static void findMid(int val) {
        if (maxHeap.isEmpty()) {
            maxHeap.add(val);
        } else {
            if (maxHeap.size() == minHeap.size()) {
                if (val <= maxHeap.peek()) maxHeap.add(val);
                else {
                    minHeap.add(val);
                    maxHeap.add(minHeap.poll());
                }
            } else if (maxHeap.size() > minHeap.size()){
                if (val <= maxHeap.peek()) {
                    maxHeap.add(val);
                    minHeap.add(maxHeap.poll());
                } else {
                    minHeap.add(val);
                }
            }
        }
        answer.append(maxHeap.peek()).append("\n");
    }
}

File: 코틀린_코테스터디/week12/가운데를말해요_1655.kt
package 코틀린_코테스터디.week12

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*


class Problem1655 {
    private var N = 0
    private var minHeap = PriorityQueue<Int>()
    private var maxHeap = PriorityQueue(Comparator.reverseOrder<Int>())
    private var answer = StringBuilder()

    fun solution() {
        val br = BufferedReader(InputStreamReader(System.`in`))
        val n = br.readLine().toInt()
        repeat(n) { findMid(br.readLine().toInt()) }
        println(answer)
    }

    private fun findMid(input: Int) {
        if (maxHeap.isEmpty()) {
            maxHeap.add(input)
        } else {
            if (maxHeap.size == minHeap.size) {
                if (input <= maxHeap.peek())
                    maxHeap.add(input)
                else {
                    minHeap.add(input)
                    maxHeap.add(minHeap.poll())
                }
            } else if (maxHeap.size > minHeap.size) {
                if (input <= maxHeap.peek()) {
                    maxHeap.add(input)
                    minHeap.add(maxHeap.poll())
                } else
                    minHeap.add(input)
            }
        }
        answer.append(maxHeap.peek()).append("\n")
    }
}

fun main() {
    Problem1655().solution()
}